homeassistant/components/co2signal/coordinator.py

```diff
@@ -1,94 +1,60 @@
         super().__init__(
             hass, _LOGGER, name=DOMAIN, update_interval=timedelta(minutes=15)
         )
-        self._entry = entry
+        self.client = client
 
     @property
     def entry_id(self) -> str:
         """Return entry ID."""
-        return self._entry.entry_id
+        return self.config_entry.entry_id
 
-    async def _async_update_data(self) -> CO2SignalResponse:
+    async def _async_update_data(self) -> CarbonIntensityResponse:
         """Fetch the latest data from the source."""
-        try:
-            data = await self.hass.async_add_executor_job(
-                get_data, self.hass, self._entry.data
-            )
-        except InvalidAuth as err:
-            raise ConfigEntryAuthFailed from err
-        except CO2Error as err:
-            raise UpdateFailed(str(err)) from err
-
-        return data
-
-
-def get_data(hass: HomeAssistant, config: Mapping[str, Any]) -> CO2SignalResponse:
-    """Get data from the API."""
-    if CONF_COUNTRY_CODE in config:
-        latitude = None
-        longitude = None
-    else:
-        latitude = config.get(CONF_LATITUDE, hass.config.latitude)
-        longitude = config.get(CONF_LONGITUDE, hass.config.longitude)
-
-    try:
-        data = CO2Signal.get_latest(
-            config[CONF_API_KEY],
-            config.get(CONF_COUNTRY_CODE),
-            latitude,
-            longitude,
-            wait=False,
-        )
-
-    except JSONDecodeError as err:
-        # raise occasional occurring json decoding errors as CO2Error so the data update coordinator retries it
-        raise CO2Error from err
-
-    except ValueError as err:
-        err_str = str(err)
-
-        if "Invalid authentication credentials" in err_str:
-            raise InvalidAuth from err
-        if "API rate limit exceeded." in err_str:
-            raise APIRatelimitExceeded from err
-
-        _LOGGER.exception("Unexpected exception")
-        raise UnknownError from err
-
-    if "error" in data:
-        raise UnknownError(data["error"])
-
-    if data.get("status") != "ok":
-        _LOGGER.exception("Unexpected response: %s", data)
-        raise UnknownError
 
-    return cast(CO2SignalResponse, data)
+        async with self.client as em:
+            try:
+                if CONF_COUNTRY_CODE in self.config_entry.data:
+                    return await em.latest_carbon_intensity_by_country_code(
+                        code=self.config_entry.data[CONF_COUNTRY_CODE]
+                    )
+
+                return await em.latest_carbon_intensity_by_coordinates(
+                    lat=self.config_entry.data.get(
+                        CONF_LATITUDE, self.hass.config.latitude
+                    ),
+                    lon=self.config_entry.data.get(
+                        CONF_LONGITUDE, self.hass.config.longitude
+                    ),
+                )
+            except InvalidToken as err:
+                raise ConfigEntryAuthFailed from err
```

Comment:
```suggestion
                raise ConfigEntryError from err
```
This would make the integration back down iirc